### DISCUSSION TITLE
Filtering stories in "update" mode

### DISCUSSION THREAD

**Discussion by bartekczyz** [lost-pixel/lost-pixel#432] [General]
Hello everyone! :)

We want to migrate from `loki` to `lost-pixel`. We use LP already for different, smaller product, and we're happy with how everything works. 

### The problem

We have a storybook with loads of stories (~640 and growing). Each one of them is rendered twice (mobile and desktop breakpoints), which gives us ~1280 stories. We also want to snapshot light / dark mode, which would be another ~640 stories, so ~1920 in total.

Right now it takes ~15 mins to take 1300 shots on M1 Pro, but that's still a lot.

### The solution

We're happy with running "partial" checks and updates on local machines and checking all of them on the CI. For local dev, we can use the [filterShot](https://github.com/lost-pixel/lost-pixel/blob/c80a096344108a78a2f762bdfcb8da9336bd0da3/src/config.ts#L364) to filter stories by name, using regexes and stuff, but it in update mode, it [removes filtered out files](https://github.com/lost-pixel/lost-pixel/blob/main/src/runner.ts#L70-L73). We want to keep the baseline files unless they're really removed from the story files.

### The proposal

I'm happy to create a PR that supports updating filtered stories only (without removing excluded by the filter):

1. add a new optional config function, let's call it `filterItemsToCheck` for now
2. update [`runner.ts`](https://github.com/lost-pixel/lost-pixel/blob/main/src/runner.ts#L66) so that it runs on filtered shot items
```ts
    const filteredShotItems = config.filterItemsToCheck
      ? shotItems.filter(config.filterItemsToCheck)
      : shotItems;
    const { aboveThresholdDifferenceItems, noBaselinesItems } =
      await checkDifferences(filteredShotItems);
```

@d-ivashchuk @chriskalmar thoughts? :)

edit: turns out it's going to be a bit more complex, because `createShots` actually do execute the shots

**Comment by bartekczyz** [comment]
Here's the PR https://github.com/lost-pixel/lost-pixel/pull/433 :)

**Reply by yashaswi1004** [reply]
@bartekczyz can you tell how to use [filterShot](https://github.com/lost-pixel/lost-pixel/blob/c80a096344108a78a2f762bdfcb8da9336bd0da3/src/config.ts#L364) to filter stories by name?

How to configure that?

and i also want to know if there is any way to run lost-pixel on certain stories from the terminal using a script instead of changing `lostpixel.config.ts`

**Reply by bartekczyz** [reply]
Hey @yashaswi1004, you can use `filterShot` by passing a function that returns a boolean:

```ts
// lostpixel.config.ts

function filterShot(shot) {
  return !!shot.shotName?.match(/* your regex */)
}

// ...

export const config: CustomProjectConfig = {
  filterShot,
}

```


but be aware that this will **remove all existing story shots if they don't match the filter**.

If you want to only check _some_ snapshots, you should use `filterItemsToCheck`. It will only check the filter items, but won't remove the ones that doesn't match the filter:

```ts
// lostpixel.config.ts

function filterItemsToCheck(shot) {
  return !!shot.shotName?.match(/* your regex */)
}

// ...

export const config: CustomProjectConfig = {
  filterItemsToCheck,
}

```

**Reply by yashaswi1004** [reply]
Thank you @bartekczyz and do you know if there is any way to run lost-pixel on certain stories/ particular component from the terminal using a script instead of changing `lostpixel.config.ts`?

**Reply by bartekczyz** [reply]
Sure, here's how I did it:

```ts
const filter = fs.existsSync('.lostpixel/filter') ? fs.readFileSync('.lostpixel/filter', 'utf8').trim() : '';
const storyFilterRegExp = new RegExp(filter, 'i');
const filterItemsToCheck = filter ? (shot) => !!shot.shotName?.match(storyFilterRegExp) : undefined;
```

This will read the filter from a file. I have a script that does some stuff, writes filter regex to a file and runs lost pixel.

Then I run it: `pnpm visual --filter "components-example-component"`. Unfortunately there's no way to pass this filter straight into the cli, you need to handle it on your own

**Reply by yashaswi1004** [reply]
Thanks a lot @bartekczyz 
i will try this
